cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -109,13 +122,142 @@ public Void call() throws Exception {
+        final Processor<LocationStatus, Set<Location>> processor = getLocationProcessor(namespaceId, tillTime);
+        try {
+          impersonator.doAs(namespaceId, new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+              Locations.processLocations(namespacedLogBaseDir, true, processor);
+              return null;
+            }
+          });
+        } catch (Exception e) {
+          LOG.warn("Got exception when deleting path {}", namespacedLogBaseDir, e);
+        }
+        Set<Location> namespaceLogFiles = processor.getResult();
+
+        LoggingContext loggingContext = new LogNamespaceLoggingContext(namespaceId.getNamespace());
+        String nextRowKey = loggingContext.getLogPartition();
+        do {
+          Processor<Location, Set<Location>> scanProcessor = getScannedFilesProcessor();
+          nextRowKey = fileMetaDataManager.scanFiles(nextRowKey, tillTime, scanProcessor);
+          for (final Location locationToDelete : Sets.difference(namespaceLogFiles, scanProcessor.getResult())) {
+            try {
+              deleteLogFiles(parentDirs, namespaceId, namespacedLogBaseDir.toString(), locationToDelete);
+            } catch (Exception e) {
+              LOG.warn("Got exception when deleting path {}", locationToDelete, e);
+            }
+            namespacedLogBaseDirMap.put(namespacedLogBaseDir.toString(), namespaceId);
+          }
+        } while (nextRowKey != null);
+      }
+    }
+  }
+
+  private Processor<Location, Set<Location>> getScannedFilesProcessor() {
+    return new Processor<Location, Set<Location>>() {
+      private Set<Location> locations = new HashSet<>();
+
+      @Override
+      public boolean process(final Location input) {
+        locations.add(input);
+        return true;
+      }
+
+      @Override
+      public Set<Location> getResult() {
+        return locations;
+      }
+    };
+  }
+
+  /**
+   * Logging context for namespace
+   */
+  public class LogNamespaceLoggingContext extends NamespaceLoggingContext {
+    /**
+     * Constructs NamespaceLoggingContext.
+     *
+     * @param namespaceId namespace id
+     */
+    public LogNamespaceLoggingContext(String namespaceId) {
+      super(namespaceId);
+    }
+  }
+
+  private Processor<LocationStatus, Set<Location>> getLocationProcessor(final NamespaceId namespaceId,
+                                                                        final long tillTime) {
+    // get list of log files older than retention duration
+    return new Processor<LocationStatus, Set<Location>>() {
+      private Set<Location> locations = new HashSet<>();
+
+      @Override
+      public boolean process(final LocationStatus input) {
+        try {
+          Location location = impersonator.doAs(namespaceId, new Callable<Location>() {
+            @Override
+            public Location call() throws Exception {
+              return rootLocationFactory.create(input.getUri());
```

Comment:
why use rootLocationFactory at some places and namespacedLocationFactory at others?